### PR TITLE
Rewrite create circuit CLI command

### PR DESCRIPTION
**Pull request by eloaverona**
This PR rewrites the `circuit create` command so it does not accept a circuit definition in a YAML file, but rather it creates the circuit definition based on CLI arguments. 

### Command to create a circuit (proposal)
```
splinter circuit create
    --node <node_alias>|<node_id>::<endpoint>
    --service <service_id>::<node_id>
    --service-type <service_id_glob>::<service_type>
    --auth-type trust
    --management <management_type>
    --service-arg <service_id_glob>::<key>=<value>
    --metadata <application_metadata>
    --metadata-encoding <`json` or `string`>
    --service-peer-group <service-id-glob-list>
```

### Details

The `--node` argument accepts either a node alias or a combination of node_id and node_endpoint. If using the node alias, the features `node-alias` must be enabled and a alias must have been set via the `node alias add` command. 

The `--service` argument is used to add a service running on a specific node. It specifies the service_id and allowed_nodes portion of the roster definiton.

The `--service-type` field takes <service_id>::<service_type>. service_id
can be a glob to match service ids. These are all valid:

  --service-type grid-scabbard-a:scabbard
  --service-type grid-scabbard-*:scabbard
  --service-type *:scabbard
  --service-type *scabbard*:scabbard

If not given, `--service-type` defaults to the value set with the 'splinter circuit default set service-type' command. If that has not been set, there is no default and an error is returned to the user. 

If not given, `--management` defaults to the value set with the 'splinter circuit default set management-type' command. If that has not been set, an error will be returned. This fills in the circuit_management_type in the circuit definition.

The auth-type can only be specified if the feature `circuit-auth-type` is enabled. If the feature is disabled or if not specified, `--auth-type` defaults to 'trust'. This is used to fill in the authorization_type filed of the circuit definition.

The --service-arg argument will fill the `arguments` field in the SplinterService definition It takes `<service_id_glob>::<key>=<value>`. The service_id_glob will indicate the services that that argument should apply to. `<key>` will indicate the key for the argument and `<value>` the value of the argument.

The `--metadata` argument will file the `application_metadata` part of the circuit definition. 

The `--metadata-encoding` argument will specify how the metadata should be encoded, it defaults to value `string`. If the value is `string`, the CLI will simply serialize the string value. If the type is `json` the CLI will build a valid json object before serializing it. 

The `--service-peer-group` argument will fill the `peer_services` argument within the 'arguments' list in a service definition.  It takes a comma-separated list of service-ids or service-id-globs. All services within a peer service group will be included in the `peer_service` argument for each service in the group

### To test 

1. Start the Gameroom example docker network:
   1. In the /splinter directory run:
`docker-compose -f examples/gameroom/docker-compose.yaml up`

2. Get the private key for alice: 
   1. In the /splinter directory run:
   ```
   docker-compose -f examples/gameroom/docker-compose.yaml run generate-key-registry bash
   cat /key_registry/alice.priv
   ```
   2. Copy the output and save it in a file in your local machine

3.  Get the public key for alice: 
   ```
   cat /key_registry/alice.pub
   ```
5. Build the splinter CLI in your local machine

4. Run the circuit create command from your local machine:
    1. In the /splinter/cli directory run:
```
cargo run --features circuit -- -vv circuit create \
    --url http://localhost:8088 \
    --key <path to alice's private key> \
    --node acme-node-000::tls://splinterd-node-acme:8044 \
    --node bubba-node-000::tls://splinterd-node-bubba:8044  \
    --service acme-scabbard::acme-node-000 \
    --service bubba-scabbard::bubba-node-000 \
    --service-type *::scabbard \
    --management gameroom \
    --service-arg *::admin_keys=<alices' pub key> \
    --service-arg bubba-scabbard::peer_services=acme-scabbard \
    --service-arg acme-scabbard::peer_services=bubba-scabbard \
    --metadata "{\"alias\":\"great_gameroom\",\"scabbard_admin_keys\":[\"alices' pub key\"]}"
```

6. Go to `localhost:8081`, log in and see that an invitation for the `great_gameroom` shows in the [invitations page](http://localhost:8081/#/dashboard/invitations). 

7. To test with node aliases and defaults set (must have `node-alias` feature enabled`:
```
cargo run --features circuit,node-alias -- -vv node alias add acme acme-node-000 tls://splinterd-node-acme:8044
cargo run --features circuit,node-alias -- -vv node alias add bubba bubba-node-000 tls://splinterd-node-bubba:8044
cargo run --features circuit,node-alias -- circuit default set service-type scabbard
cargo run --features circuit,node-alias -- circuit default set management-type gameroom

cargo run --features circuit,node-alias -- -vv circuit create \
    --url http://localhost:8088 \
    --key <path to alice's private key> \
    --node acme \
    --node bubba \
    --service acme-scabbard::acme-node-000 \
    --service bubba-scabbard::bubba-node-000 \
    --service-arg *::admin_keys=<alices' pub key> \
    --service-arg bubba-scabbard::peer_services=acme-scabbard \
    --service-arg acme-scabbard::peer_services=bubba-scabbard \
    --metadata "{\"alias\":\"great_gameroom\",\"scabbard_admin_keys\":[\"alices' pub key\"]}"
```

8. With `--service-peer-group`:
```
cargo run --features circuit,node-alias --  -vv circuit create \
    --url http://localhost:8088 \
    --key <path to alice's private key> \
    --node acme \
    --node bubba \
    --service acme-scabbard::acme-node-000 \
    --service bubba-scabbard::bubba-node-000 \
    --service-arg *::admin_keys=<alices' pub key> \
    --service-peer-group acme-scabbard,bubba-scabbard \
    --metadata "{\"alias\":\"great_gameroom\",\"scabbard_admin_keys\":[\"alices' pub key\"]}"
```
8. With `--metada-encoding` set to json:
```
cargo run --features circuit,node-alias --  -vv circuit create \
    --url http://localhost:8088 \
    --key <path to alice's private key> \
    --node acme \
    --node bubba \
    --service acme-scabbard::acme-node-000 \
    --service bubba-scabbard::bubba-node-000 \
    --service-arg *::admin_keys=$(cat alice.pub) \
    --service-peer-group acme-scabbard,bubba-scabbard \
    --metadata-encoding json \
    --metadata alias=great_gameroom \
    --metadata scabbard_admin_keys=[\"$(cat alice.pub)\"]
```

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -192,10 +192,68 @@ fn run() -> Result<(), CliError> {
...
+                            Arg::with_name("node")
+                                .long("node")
                                 .takes_value(true)
                                 .required(true)
-                                .help("Path to a yaml file that defines the circuit proposal"),
+                                .multiple(true)
+                                .help(
+                                    "Node that are part of the circuit. \
+                                  Format: <node_id>::<endpoint>. \
+                                  Endpoint is optional is node alias has been set.",
```
Comment (agunde406) [inline cli/src/main.rs]: Endpoint is optional is node alias has been set -> Endpoint is optional if node alias has been set

File: cli/src/main.rs [hunk 2]
```diff
@@ -192,10 +192,68 @@ fn run() -> Result<(), CliError> {
...
+                                .help(
+                                    "Special arguments to be passed to the service. \
+                                Format <service_id>::<key>=<value>",
+                                ),
+                        )
+                        .arg(
+                            Arg::with_name("management_type")
+                                .long("management-type")
+                                .takes_value(true)
+                                .help("Management type for a service"),
```
Comment (agunde406) [inline cli/src/main.rs]: management type is for the circuit, not a service

File: cli/src/main.rs [hunk 3]
```diff
@@ -192,10 +192,68 @@ fn run() -> Result<(), CliError> {
...
+                        )
+                        .arg(
+                            Arg::with_name("authorization_type")
+                                .long("auth-type")
+                                .possible_values(&["trust"])
+                                .default_value("trust")
+                                .takes_value(true)
+                                .help("Authorization type for a service"),
```
Comment (agunde406) [inline cli/src/main.rs]: auth type is for the circuit, not a service

**Comment by peterschwarz** [comment]
The example for testing should have `--key` instead of `--private-key-file`

**Comment by peterschwarz** [comment]
Does gameroom need to be compiled with experimental features?

File: cli/src/main.rs (hunk 2 above)
Comment (eloaverona) [reply cli/src/main.rs]: Updated

File: cli/src/main.rs (hunk 3 above)
Comment (eloaverona) [reply cli/src/main.rs]: Updated

File: cli/src/main.rs (hunk 1 above)
Comment (eloaverona) [reply cli/src/main.rs]: updated

**Comment by eloaverona** [comment]
> The example for testing should have --key instead of --private-key-file
Fixed

> Does gameroom need to be compiled with experimental features?

No, are you having any issues?

File: cli/src/main.rs [hunk 4]
```diff
@@ -192,10 +192,68 @@ fn run() -> Result<(), CliError> {
...
                         )
                         .arg(
-                            Arg::with_name("path")
+                            Arg::with_name("node")
+                                .long("node")
                                 .takes_value(true)
                                 .required(true)
-                                .help("Path to a yaml file that defines the circuit proposal"),
+                                .multiple(true)
+                                .help(
```
Comment (agunde406) [inline cli/src/main.rs]: Please change this to long_help for better help formatting, same for all help that has multiple lines

**Comment by vaporos** [comment]
circuit default with no args returns an error:

% cargo run --features=experimental -- circuit default
    Finished dev [unoptimized + debuginfo] target(s) in 0.22s
     Running `/Users/amundson/projects/review/splinter-pr-492/target/debug/splinter circuit default`
ERROR: received invalid subcommand

**Comment by vaporos** [comment]
Also, please shorten "--management-type" to "--management"

File: cli/src/main.rs [hunk 5]
```diff
@@ -246,7 +246,17 @@ fn run() -> Result<(), CliError> {
...
                     .help("Application metadata for the circuit proposal"),
+            )
+            .arg(
+                Arg::with_name("metadata_encode")
+                    .long("metadata-encode")
+                    .takes_value(true)
+                    .possible_values(&["json", "string"])
+                    .default_value("string")
+                    .requires("application_metadata")
+                    .help("Set the enconding for the application metadata"),
```
Comment (rbuysse) [inline cli/src/main.rs]: enconding -> encoding

File: cli/src/main.rs [hunk 6]
```diff
@@ -246,7 +246,17 @@ fn run() -> Result<(), CliError> {
                 Arg::with_name("application_metadata")
                     .long("metadata")
                     .takes_value(true)
+                    .multiple(true)
                     .help("Application metadata for the circuit proposal"),
+            )
+            .arg(
+                Arg::with_name("metadata_encode")
+                    .long("metadata-encode")
```
Comment (rbuysse) [inline cli/src/main.rs]: should this be metadata-encoding?

Comment (eloaverona) [reply cli/src/main.rs]: updated

File: cli/src/main.rs (hunk 5 above)
Comment (eloaverona) [reply cli/src/main.rs]: updated

**Pull request merged**